### ISSUE TITLE
Modem support

### ISSUE THREAD

**Issue by xc-racer99** [PabloPL/linux#23]
Branch created called not-for-upstream/modem

Has been tested on Fascinate4G with https://git.replicant.us/replicant/external_libsamsung-ipc/tree/?h=replicant-6.0 and the patches from https://github.com/xc-racer99/android_patches/blob/8e410ce0996bb3170865a4b7ea375726bf804c1c/external_libsamsung-ipc.patch applied

Modem boots and registers on network.  Requires the modem.bin to be in /radio/modem.bin plus that the EFS is present at /efs

**Comment by PabloPL** [comment]
https://github.com/fourkbomb/linux/tree/modem
There is some rewritten samsung ipc driver for mainline. We would need to implement onedram driver.

**Comment by xc-racer99** [comment]
Do you really think it's worthwhile trying to shoehorn the onedram bits into the i9300 driver?  I suppose we'd need an equivalent of https://github.com/fourkbomb/linux/commit/459dd2c610c44d90683953f8e9dcb68df2c91807 for onedram.  The branch I created uses the Nexus S (crespo) driver with forward-ported to 4.20 plus support for the modem in the Fascinate 4G.

Do you know if https://github.com/fourkbomb/linux/tree/modem was ever submitted to mainline and what the reaction was?

**Comment by PabloPL** [comment]
Probably it wasn't submited to upstream.

At first it would be great to have old driver working (like You did), but later i think we could try to make it work with fourkbomb driver (since it was designed for mainline, not for 2.6/3.0 kernels) - having old, working code as base would make it easier to develop (like freedreno/lima/panfrost/etc are doing).
But it's up to You, this is just my sugestion ;).

**Comment by xc-racer99** [comment]
Ok, sounds good, I think I'll leave it like it is for now as there's no guarantee that fourkbomb's driver will ever see mainline either.  The issue with fourkbomb's driver is that you still need a separate modem control driver for power states such as https://github.com/fourkbomb/linux/commit/913bdb1ad530a3c32f9d0ae3aef5c9afaabfd567

The fact that it works as-is with userspace is one of the reasons I chose it.  To me, whether or not it makes it into mainline is rather beside the point.

And the main reason I worked on this at all was so that I could return to the audio driver and make sure I've actually setup the clocking for the modem part correctly.

**Comment by xc-racer99** [comment]
It should also be noted that there are two 2.6.35/3.0 onedram/modemctl drivers - one that is used on the SGS series, one that is used for the Nexus S.  I chose the Nexus S driver as it is much, much simpler plus has a full history of commits - it was worked on primarily by Google and not by Samsung.  I also didn't bother choosing the CDMA commits as there's no free userspace anyways.

**Comment by PabloPL** [comment]
Yes, Nexus S has much better drivers (looking at code) than other aries devices.

**Comment by PabloPL** [comment]
libsamsung-ipc has switched to gpl2 some time ago (that was preventing it from being supported by ofono).
https://lists.ofono.org/pipermail/ofono/2012-September/013777.html
Maybe we could update those patches and send them again.

**Comment by xc-racer99** [comment]
>  Maybe we could update those patches and send them again.

Unfortunately, libsamsung-ipc has also undergone a fairly substantial rewrite since then.  There is a WIP set for libsamsung-ipc v0.3 at https://github.com/fourkbomb/ofono

I have no experience with Glib, so I'm not sure if I could do this at this point.  Additionally, my device needs the RFS layer setup before anything else, so I would need to implement that as well.

**Comment by xc-racer99** [comment]
Note that the libsamsung-ipc ofono patches have external projects located at https://github.com/morphis/samsung-rfs-mgr and https://github.com/morphis/samsung-modem-mgr

I'm working my way through rebasing the original ofono patches on the latest master, fixing the build as I go along.  Work is at https://github.com/xc-racer99/ofono but it untested as of writing.

**Comment by xc-racer99** [comment]
Ok, I now have all the ofono patches building.  Unfortunately, there is zero chance it works.  Previously, you could get the fd that the libsamsung-ipc driver was internally using its "poll" function, but you can no longer do so.  The ofono patches relied on this (using g_io_channel_unix_new() which was then listened on).  Since this is no longer the case, there's a few things we could try:
1) Create a separate daemon that is responsible for direct interaction with the libsamsung-ipc.  It would create a socket and the ofono code would connect to this socket.
2) Create a custom GIOChannel wrapper around the libsamsung-ipc poll function
3) Re-add some sort of fd listening to libsamsung-ipc
4) Create a separate thread inside ofono for one of the above

Considering I've never done any GLib programming, this might be beyond me.  I think I'll try implementing option 2 first as it would involve the least amount of changes to the pre-existing patches.

**Comment by herbsmn** [comment]
Have you looked at the oFono-based Java RIL for Android that Replicant is planning to use? https://github.com/scintill/android_frameworks_opt_telephony_ril_ofono

There are many Galaxy S3 versions that use Qualcomm CPUs, but Replicant is only working with the S3 versions that use the Exynos SoC.

GNUtoo said that there are plans to upstream the Samsung IPC driver and that the first driver will probably be the firmware loading code.

I think that this is where you can follow GNUtoo's progress: https://git.replicant.us/contrib/GNUtoo/hardware_replicant_libsamsung-ipc/log/?h=GNUtoo/modem-i9300

**Comment by xc-racer99** [comment]
> Have you looked at the oFono-based Java RIL for Android that Replicant is planning to use? https://github.com/scintill/android_frameworks_opt_telephony_ril_ofono

I've noticed it was around, but haven't really looked at it at all.  I'm more interested in the pmOS/more traditional linux distro side of things than Android right now which doesn't need the Java or RIL layer.  I've been using the tests that ship with ofono for testing things.


> There are many Galaxy S3 versions that use Qualcomm CPUs, but Replicant is only working with the S3 versions that use the Exynos SoC.
> 
> GNUtoo said that there are plans to upstream the Samsung IPC driver and that the first driver will probably be the firmware loading code.
> 
> I think that this is where you can follow GNUtoo's progress: https://git.replicant.us/contrib/GNUtoo/hardware_replicant_libsamsung-ipc/log/?h=GNUtoo/modem-i9300

Good to know I've been watching in the right place :)  If we can have a proper firmware loading code abstraction in the kernel, that would be great.  Right now, the two (known) modems used in the first-gen Galaxy S series (XMM6160 and STE M5730) use different initialization procedures but substantially the same communication method once they're booted.
